packages/labs/eleventy-plugin-lit/demo/_includes/default.html

```diff
@@ -1,12 +1,56 @@
 <!DOCTYPE html>
-<html lang="en">
+<html>
   <head>
-    <meta charset="utf-8" />
-    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
-    <title>{{ title }}</title>
+    <!-- As an optimization, immediately begin fetching the JavaScript modules
+        that we know for sure we'll eventually need. It's important we don't
+        execute them yet, though. -->
+    <link rel="modulepreload" href="./_js/components.js" />
   </head>
+
   <body>
+    <!-- Pre-rendered Lit components will be generated here. -->
     {{ content }}
+
+    <!-- At this point, browsers with native shadow DOM support will already
+         be able to paint the initial fully styled state your components,
+         without executing a single line of JavaScript! However, the components
+         aren't interactive yet -- that's what hydration is for. -->
+
+    <!-- Use a type=module script so that we can use dynamic module imports.
+         Note this pattern will not work in IE11. -->
+    <script type="module">
+      (async () => {
+        // Start fetching the Lit hydration support module (note the absence
+        // of "await" -- we don't want to block yet).
+        const litHydrateSupportInstalled = import(
+          'lit/experimental-hydrate-support.js'
+        );
+
+        // Check if we require the declarative shadow DOM ponyfill. As of
+        // February 2022, Chrome and Edge have native support, but Firefox
+        // and Safari don't yet.
+        if (!HTMLTemplateElement.prototype.hasOwnProperty('shadowRoot')) {
+          // Fetch the declarative shadow DOM ponyfill.
+          const {hydrateShadowRoots} = await import(
+            '@webcomponents/template-shadowroot/template-shadowroot.js'
```

Comment:
Not sure there's a way to avoid this, but I assume if you slow your network down on e.g. Safari/Firefox, you'd get fouc because there's no way to conditionally load this _and_ make it run sync before paint?

FYI if you haven't seen it, there's a mostly untested [mutation-observer](https://github.com/webcomponents/template-shadowroot/blob/master/src/_implementation/mutation_observer.ts) implementation of the polyfill that could make this more robust by loading it up front and letting it upgrade as the parser yields (at the cost of loading it unconditionally, even on Chrome). We wanted to perf-test all the options and figure out the most robust/fastest way to polyfill it (which is why it's in such a manual state right now), so this is probably another thing to put on the roadmap.

Reply:
Yeah good point, even on my machine there is a FOUC on Firefox with the demo. I also was thinking of a pattern like:

```js
<style>
  body[dsd-pending] { display: none; }
</style>
<body dsd-pending>
  <script>
  if (HTMLTemplateElement.prototype.hasOwnProperty('shadowRoot')) {
    body.removeAttribute('dsd-pending');
  }
  </script>

  {{content}}

  <script type="module">
    ...
    if (!HTMLTemplateElement.prototype.hasOwnProperty('shadowRoot')) {
      ...
      hydrateShadowRoots(document.body);
      body.removeAttribute('dsd-pending');
    }
    ...
  </script>
</body>
```

The downside of this is that it wouldn't allow for any incremental painting before all DOM content is parsed, and it's a little complicated.

The `MutationObserver` implementation seems worth pursuing though, yeah. If it's small, inlining it in the `<head>` could turn out to be pretty optimal.

Reply:
Filed https://github.com/lit/lit/issues/2490 and added it to the Roadmap section.

Reply:
I like that pattern, it's very much like the old Polymer feature or `:not(:defined) { display: none; }`

There's also another ponyfill approach where the JS isloaded up front and we emit calls to attach shadow roots in inline scripts right after each template.